src/rust/engine/src/externs/interface.rs

```diff
@@ -830,6 +840,40 @@ fn nailgun_server_await_shutdown(py: Python, nailgun_server_ptr: PyNailgunServer
   })
 }
 
+fn strongly_connected_components(
+  py: Python,
+  adjacency_lists: Vec<(PyObject, Vec<PyObject>)>,
+) -> CPyResult<Vec<Vec<PyObject>>> {
+  let mut graph: DiGraph<Key, (), u32> = Graph::new();
+  let mut node_ids: HashMap<Key, _> = HashMap::new();
+
+  for (node, adjacency_list) in adjacency_lists {
+    let node_key = externs::key_for(node.clone_ref(py).into())?;
+    let node_id = *node_ids
+      .entry(node_key)
+      .or_insert_with(|| graph.add_node(node_key));
+    for dependency in adjacency_list {
+      let dependency_key = externs::key_for(dependency.clone_ref(py).into())?;
+      let dependency_id = node_ids
+        .entry(dependency_key)
+        .or_insert_with(|| graph.add_node(dependency_key));
+      graph.add_edge(node_id, *dependency_id, ());
+    }
+  }
+
+  Ok(
+    petgraph::algo::tarjan_scc(&graph)
```

Comment:
I love great libraries like this :)

Reply:
Maybe add a brief comment motivating why we use Tarjan instead of Kosaraju, given that both are offered:

https://www.quora.com/Is-there-any-difference-or-advantages-or-disadvantages-between-Kosarajus-algorithm-and-Tarjans-algorithm-for-finding-strongly-connected-component-Which-one-is-most-useful?share=1

https://www.geeksforgeeks.org/comparision-between-tarjans-and-kosarajus-algorithm/